add new blog post pause a container (#74)
metalbear-co/metalbear.co: content/en/blog/pause-a-container/index.md

---

> title: "Pause a Container - How we built, and why we deprecated, our container pause feature"

Reviewer: ```suggestion
title: "On Pausing Containers - how we built, and why we deprecated, our container pause feature"
```

description: "How we implemented the container pause feature, and why we decided to deprecate it"
lead: "How we implemented the container pause feature, and why we decided to deprecate it"
tags:
  - cri
  - container
  - runtime
  - pause
  - mirrord
date: 2024-03-04T06:00:00+00:00
lastmod: 2024-03-04T06:00:00+00:00
draft: false
weight: 50
images: ["thumbnail.webp"]
contributors: ["Aviram Hassan"]
---

Did you know that you can ”freeze” running containers? We found out about this feature while developing a particularly handy capability for mirrord. 
mirrord lets developers run local code in the context of a remote Kubernetes pod. This lets them test their code in cloud conditions without going through CI and deployment. In most cases, stealing or mirroring incoming traffic from the remote pod to the local process is good enough. However, when you have applications that consume tasks and do more than react to incoming requests, the user might want to [pause the remote application completely](https://github.com/metalbear-co/mirrord/discussions/2016), so that it doesn’t compete with their local process for tasks.

## How Do We Pause? 

> We found out that there is a Docker API [for pausing a container](https://docs.docker.com/engine/api/v1.44/#tag/Container/operation/ContainerPause), which can be used to suspend all processes in a container. Luckily, we already work with CRI (Container Runtime Interface) API to collect information about the container, so using it for pausing is easy.

Reviewer: ```suggestion
We found out that there is a Docker API [for pausing a container](https://docs.docker.com/engine/api/v1.44/#tag/Container/operation/ContainerPause), which can be used to suspend all processes in a container. Luckily, we already work with the CRI (Container Runtime Interface) API to collect information about the container, so using it for pausing is easy.
```

The second issue we ran into was that if there were live incoming connections, those would also take time to drop, or sometimes never drop at all. We mostly saw this in gRPC services. The common case was users trying to use mirrord with pause on a service in their cluster that received gRPC connections from another microservice. The calling microservice would stay connected to the paused service rather than dropping the connection and connecting to our local process instead.

We tried finding ways to flush all connections when using pause.It seems that the Linux kernel really likes holding those connections, as even if we conntrack flush, it’d just bring back the entries on its own. The only solution we saw was a kernel module to kill these connections, but that would have been violent and ugly, plus we didn’t want to maintain a kernel module.

> The last and biggest issue we encountered was when we tried to migrate mirrord to using [Ephemeral Containers](https://metalbear.co/blog/getting-started-with-ephemeral-containers/) for the mirrord agent. The mirrord agent is a component that runs in the cluster and helps mirrord proxy data back and forth between the local process and remote pod. By default we use a Job to run the agent, but Ephemeral Containers, when available, are a better functional fit. However, it means we lose access to the container runtime API, and so we couldn’t use it to pause the container

Reviewer 2: ```suggestion
The last and biggest issue we encountered was when we tried to migrate mirrord to using [Ephemeral Containers](https://metalbear.co/blog/getting-started-with-ephemeral-containers/) for the mirrord agent. The mirrord agent is a component that runs in the cluster and helps mirrord proxy data back and forth between the local process and remote pod. By default we use a Job to run the agent, but Ephemeral Containers, when available, are a better functional fit. However, it means we lose access to the container runtime API, and so we couldn’t use it to pause the container.
```
Dot

The way it works:
The mirrord Operator creates a new pod that is part of the deployment, copying the configmap, mount, sidecars, and anything that can be useful.
The Operator scales down the deployment, shutting down the rest of pods for the duration of the mirrord session.
Given the delicate state it can leave the cluster in, we decided to have this feature as part of mirrord for Teams. The general mandate is that anything that is stateless and won’t break the cluster can be in the OSS, while anything that manages state should be done by the Operator.

## Pros and Cons of the New Solution

**Pros**:
1. Immediate effect both for queues and incoming traffic.
2. It uses a stable, high-level k8s API.

**Cons**:
1. We don’t use the original base image for the pod replica, so if there are files in the image base, those aren’t available to the local process
2. Following the previous point, we don’t run the original Docker entrypoint, which might be an issue if it contains meaningful logic (for example, we saw users that copy configuration files from configmap to other paths on start)
Note that both issues have workarounds, but they’re still a bit of a hassle.

## Final Notes

> It’s no fun removing a feature you’ve worked hard on, but it’s even worse to leave broken functionality in your product, and so we’ll be deprecating the pause feature in the coming months. Hopefully, at least the lessons we learned while working on it and documented in this blog post would be useful or interesting for some of you. And if you want to try out copy/scaledown functionality, check out [mirrord for Teams](http://app.metalbear.co).

Reviewer: ```suggestion
It’s no fun removing a feature you’ve worked hard on, but it’s even worse to leave broken functionality in your product, and so we’ll be deprecating the pause feature in the coming months. Hopefully, at least the lessons we learned while working on it and documented in this blog post would be useful or interesting for some of you. And if you want to try out copy/scaledown functionality, check out [mirrord for Teams](http://app.metalbear.co)!
```
